[FC] Eagerly fetch chunks in the cow store

maggie-lou opened the pull request:
If we access a snapshot chunk, use a background queue to eagerly fetch the next X contiguous chunks, anticipating that they will also be accessed.

**Some performance results:**
The performance tests build the buildbuddy repo from a snapshot that is cached in the remote cache, but nothing is cached locally in filecache. Every testing scenario was run 6 times. 

Baseline: (45.8s - 50.6s)
Best config - 8 concurrent threads, max 1000 fetches per sec: (24.9s - 27.2s)

**Some additional sample tests for context:**

4 concurrent threads, max 1000 fetches per sec: (25.7s - 27.6s)
4 concurrent threads, max 500 fetches per sec: (28.5 - 29.5s)
8 concurrent threads, max 100 fetches per sec: (37.5s - 40.2)
-  Additional threads not that helpful, because rate limiter is so low
4 concurrent threads, max 100 fetches per sec: (38.1 - 42.3)
4 concurrent threads, max 1000 per sec, 10 contiguous chunks: (27.5 - 30)
- All other tests are with 4 contiguous chunks
- Fetching more chunks each time leads to a slowdown in performance

maggie-lou commented:
Current dependencies on/for this PR:
* master
  * **PR #5076** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/5076" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/5076?utm_source=stack-comment).

bduffany reviewed (commented):
This implementation looks great.

Before submitting, I think it'd be good to do some performance testing, possibly with different tuning values (e.g. N=0 vs N=4 / concurrency=4 vs concurrency=8, etc.). Otherwise we're a bit blind as to whether the eager fetching provides a net benefit, or how things will change if we try to adjust these tuning values.

bduffany commented on enterprise/server/remote_execution/copy_on_write/copy_on_write.go:
```diff
@@ -473,6 +516,73 @@ func (s *COWStore) initDirtyChunk(offset int64, size int64) (ogChunk *Mmap, newC
+	if c == nil {
+		return nil
+	}
+
```
nit: maybe move the rateLimiter.Wait() after this line?

maggie-lou replied:
I'm wondering if the rate limiter could also help protect over-locking on `chunks`. I'm just going to leave it as is for now - probably wouldn't be a big difference either way

bduffany commented:
As a rough performance test, it might be useful to try running `enterprise/tools/vmstart` now that it supports remote snapshots. Some behavior I noticed was that the first LoadSnapshot took a long time (~30s), then each subsequent one got faster

maggie-lou commented:
> This implementation looks great.
> 
> Before submitting, I think it'd be good to do some performance testing, possibly with different tuning values (e.g. N=0 vs N=4 / concurrency=4 vs concurrency=8, etc.). Otherwise we're a bit blind as to whether the eager fetching provides a net benefit, or how things will change if we try to adjust these tuning values.

I added some performance test results to the pr body and updated the config to match the results. Also made them configurable by flag
